instrumentation/opentelemetry-instrumentation-asyncio/README.rst

```diff
@@ -0,0 +1,130 @@
+OpenTelemetry asyncio Instrumentation
+======================================
+
+|pypi|
+
+.. |pypi| image:: https://badge.fury.io/py/opentelemetry-instrumentation-asyncio.svg
+   :target: https://pypi.org/project/opentelemetry-instrumentation-asyncio/
+
+AsyncioInstrumentor: Tracing Requests Made by the Asyncio Library
+
+
+The opentelemetry-instrumentation-asycnio package allows tracing asyncio applications.
+The metric for coroutine, future, is generated even if there is no setting to generate a span.
+
+
+Set the name of the coroutine you want to trace.
+-------------------------------------------------
+.. code:: bash
+
+    export OTEL_PYTHON_ASYNCIO_COROUTINE_NAMES_TO_TRACE=coro_name,coro_name2,coro_name3
+
+If you want to keep track of which function to use in the to_thread function of asyncio, set the name of the function.
+------------------------------------------------------------------------------------------------------------------------
+.. code:: bash
+
+    export OTEL_PYTHON_ASYNCIO_TO_THREAD_FUNCTION_NAMES_TO_TRACE=func_name,func_name2,func_name3
+
+For future, set it up like this
+-----------------------------------------------
+.. code:: bash
+
+    export OTEL_PYTHON_ASYNCIO_FUTURE_TRACE_ENABLED=true
+
+Run instrumented application
+-----------------------------
+1. coroutine
+--------------------
+.. code:: python
+
+    # export OTEL_PYTHON_ASYNCIO_COROUTINE_NAMES_TO_TRACE=sleep
+
+    import asyncio
+    from opentelemetry.instrumentation.asyncio import AsyncioInstrumentor
+
+    AsyncioInstrumentor().instrument()
+
+    async def main():
+        await asyncio.create_task(asyncio.sleep(0.1))
+
+    asyncio.run(main())
+
+2. future
+--------------------
+.. code:: python
+
+    # export OTEL_PYTHON_ASYNCIO_FUTURE_TRACE_ENABLED=true
+
+    loop = asyncio.get_event_loop()
+
+    future = asyncio.Future()
+    future.set_result(1)
+    task = asyncio.ensure_future(future)
+    loop.run_until_complete(task)
+
+3. to_thread
+--------------------
+.. code:: python
+
+    # export OTEL_PYTHON_ASYNCIO_TO_THREAD_FUNCTION_NAMES_TO_TRACE=func
+
+    import asyncio
+    from opentelemetry.instrumentation.asyncio import AsyncioInstrumentor
+
+    AsyncioInstrumentor().instrument()
+
+    async def main():
+        await asyncio.to_thread(func)
+
+    def func():
+        pass
+
+    asyncio.run(main())
+
+
+asyncio metric types
+----------------------
+
+* `asyncio.futures.duration` (ms) - Duration of the future
+* `asyncio.futures.exceptions` (count) - Number of exceptions raised by the future
+* `asyncio.futures.cancelled` (count) - Number of futures cancelled
+* `asyncio.futures.created` (count) - Number of futures created
+* `asyncio.futures.active` (count) - Number of futures active
+* `asyncio.futures.finished` (count) - Number of futures finished
+* `asyncio.futures.timeouts` (count) - Number of futures timed out
```

Comment:
Can at least some of these metrics be collapsed into a single metric with an attribute indicating the "state"? As an example see `system.cpu.time` metric which has a `system.cpu.state` attribute https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemcputime

Keep in mind that the sum across dimensions should be meaningful which I'm not sure is the case here. Something like this:

- `asyncio.futures.count` of type Counter
  - `state`: `cancelled | finished | timed_out | exception`
  - note that sum across all of these is total number created/finished so you don't need that metric
- `asyncio.futures.active` of type UpDownCounter

The same comment applies to the below metrics

Reply:
I've made the changes you suggested.
I've learned a lot from you today, thanks for the feedback.